#include<stdio.h>
void usage(){
	fprintf(stderr, "rasqual:\n");
	fprintf(stderr, "  RASQUAL (Robust Allele Specific Quantitation and quality controL) maps QTLs\n");
	fprintf(stderr, "  using population and AS signals for any sequenced based cellular tratis.\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Usage:\n");
	fprintf(stderr, "  VCF as stdin | ./rasqual [OPTIONS] -y <counts.bin> -k <offsets.bin> -n <sample_size> -j <feature_id> \\\n");
	fprintf(stderr, "                                     -l <#tested_SNPs> -m <#feature_SNPs> -s <feature_start(s)> -e <feature_end(s)>\n");
	fprintf(stderr, "									 \n");
	fprintf(stderr, "Output:\n");
	fprintf(stderr, "  Tested SNPs in stdout and posterior genotype in VCF file\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Mandatory arguments:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -y/--feature-counts            <string>  File name of total feature counts in binary double\n");
	fprintf(stderr, "  -k/--sample-offsets            <string>  File name of sample specific offset terms in binary double\n");
	fprintf(stderr, "  -s/--feature-starts            <string>  Feature start position(s) in comma separated value (e.g., 100,300,500)\n");
	fprintf(stderr, "  -e/--feature-ends              <string>  Feature end position(s) in comma separated value (e.g., 150,350,450)\n");
	fprintf(stderr, "  -j/--feature-id                <int>     Row number of target feature in original count table\n");
	fprintf(stderr, "  -l/--number-of-testing-snps    <int>     Number of testing SNPs in VCF (candidates of a rSNP)\n");
	fprintf(stderr, "  -m/--number-of-fsnps           <int>     Number of feature SNPs (fSNPs) in VCF\n");
	fprintf(stderr, "  -n/--sample-size               <int>     Sample size\n");
	fprintf(stderr, "  \n");
	fprintf(stderr, "Output options:\n");
	fprintf(stderr, "  \n");
	fprintf(stderr, "  -t/--lead-snp                            Output only the lead QTL SNP\n");
	fprintf(stderr, "  -i/--imprint                             Output imprinted (or randomly allelicly inactivated) feature result\n");
	fprintf(stderr, "  -g/--posterior-genotype        <string>  Output posterior genotypes as VCF format\n");
	fprintf(stderr, "  -f/--feature-name              <string>  Feature name printed on the result (default: feature ID number)\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Test SNP (rSNP) filter options:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -c/--cis-midpoint              <int>     Midpoint of cis regulatory window, e.g., TSS (default: midpoint of feature start and end)\n");
	fprintf(stderr, "  -w/--cis-window-size           <int>     Cis regulatory window size centered at the midpoint (default: 54e7)\n");
	fprintf(stderr, "  -a/--minor-allele-frequency    <double>  Minor allele frequency cutoff (default: 0.05)\n");
	fprintf(stderr, "  -h/--hardy-weinberg-pvalue     <double>  Hardy-Weinberg equiliblium P-value cutoff (default: 1.0e-8)\n");
	fprintf(stderr, "  -q/--imputation-quality        <double>  Genome imputation quality score cutoff, e.g., R^2 (MaCH/Beagle) or I^2 (IMPUTE) (default: 0.7)\n");
	fprintf(stderr, "  --rsnp                         <string>  Test only specific SNP (variant ID given in the VCF file)\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "fSNP filter options:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  --minor-allele-frequency-fsnp  <double>  Minor allele frequency cutoff for fSNPs (default: 0.0)\n");
	fprintf(stderr, "  --hardy-weinberg-pvalue-fsnp   <double>  Hardy-Weinberg equiliblium P-value cutoff for fSNPs (default: 0.0)\n");
	fprintf(stderr, "  --imputation-quality-fsnp      <double>  Genome imputation quality score cutoff for fSNPs (default: 0.0)\n");
	fprintf(stderr, "  --min-coverage-depth           <double>  Minimum depth to filter out low coverage fSNPs (default: 0.05)\n");
	fprintf(stderr, "  --effective-feature-length     <int>     Effective length of feature to calculate average depth (default: length of union features)\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Model fitting options:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -x/--covariates                <string>  File name of covariates in bianary double\n");
	fprintf(stderr, "  -p/--number-of-covariates      <int>     Number of covariates used for adjustment\n");
	fprintf(stderr, "  -r/--random-permutation                  Randomly permute samples\n");
	fprintf(stderr, "  -k2/--conditional-snps         <string>  Conditional SNP(s) ID with effect size(s) connected by colon (:) (e.g., rs1:0.12,rs2:0.34) \n");
	fprintf(stderr, "  --population-only                        Only total feature counts are modelled\n");
	fprintf(stderr, "  --as-only                                Only AS counts are modelled\n");
	fprintf(stderr, "  --null                                   Fitting null model only\n");
	fprintf(stderr, "  --force                                  Fitting model for the large numbers of fSNPs and rSNPs (#fSNPs x #rSNPs < 30,000 w/o this option)\n");
	fprintf(stderr, "  --fix-delta                              Fix sequencing/mapping error parameter (Delta=0.01)\n");
	fprintf(stderr, "  --fix-phi                                Fix reference allele mapping parameter (Phi=0.5)\n");
	fprintf(stderr, "  --fix-theta                              Fix overdispersion parameter (Theta=10000)\n");
	fprintf(stderr, "  --number-of-significant-loci   <int>     Average number of SNPs tested by joint model (default: number of tested SNPs)\n");
	fprintf(stderr, "  --n-threads                    <int>     The number of threads used for mapping QTLs in parallel\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Genotype uncertainty options:\n");
	fprintf(stderr, "  \n");
	fprintf(stderr, "  -z/--convert-imputation-score            Convert genome imputation quality score (R^2 or I^2) into allelic probability (AP)\n");
	fprintf(stderr, "  --nominal-allelic-probability            Use nomial AP (default: AP is truncated in [0.001, 0.999] to allow uncertainty)\n");
	fprintf(stderr, "  --genotype-dosage                        Use genotype dosage as AP\n");
	fprintf(stderr, "  --population-allele-frequency            Use population allele frequency as AP\n");
	fprintf(stderr, "  --fix-genotype                           Use fix genotype (in GT field)\n");
	fprintf(stderr, "  --no-posterior-update                    Not updating posterior probability\n");
	fprintf(stderr, "  \n");
	fprintf(stderr, "Other options:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -v/-V   Print verbose output (for debugging). \n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Hyperparameter options (modification unrecomended):\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -SIGMA  <double>    Variance parameter for Normal gamma distribution on Lambda (default: 10000)\n");
	fprintf(stderr, "  -BETA   <double>    Mean parameter for Normal-gamma distribution on Lambda (default: sample mean)\n");
	fprintf(stderr, "  -ABPI   <double>    Shape parameters (identical) for Beta distribution on Pi (default:10)\n");
	fprintf(stderr, "  -ABPHI  <double>    Shape parameters (identical) for Beta distribution on Phi (default:10)\n");
	fprintf(stderr, "  -ADELTA <double>    1st shape parameter for Beta distribution on Delta (default: 1.01)\n");
	fprintf(stderr, "  -BDELTA <double>    2nd shape parameter for Beta distribution on Delta (default: 1.99) \n");
	fprintf(stderr, "  -KAPPA  <double>    Shape parameter for Gamma distribution on Theta (default: 2.02)\n");
	fprintf(stderr, "  -OMEGA  <double>    Rate parameter for Gamma distribution on Theta (default: 0.2)\n");
	fprintf(stderr, "\n");
}
